File: haskell/question.hs
import System.IO (stdout, hFlush)
import Data.List (intercalate)

-- | implementation of the question function in haskell
question :: String -> [String] -> IO String
question prompt [] = do
    putStrLn prompt
    putStr ": "
    hFlush stdout
    getLine
question prompt valid = do
    putStrLn prompt
    putStr $ "(" ++ intercalate ", " valid ++ "): "
    hFlush stdout
    input <- getLine
    if input `elem` valid then return input
    else do
        putStrLn $ "\"" ++ input ++ "\" is not a valid answer!"
        question prompt valid

main = question "foo" ["bar", "baz"]

File: js/deno/question.ts
// implementation of the question function in deno
async function question(prompt: string, valid?: string[]) {
  const joinedPrompt = valid ? `${prompt}(${valid.join(", ")}):` : `${prompt}:`;

  while (true) {
    const input = globalThis.prompt(joinedPrompt)?.trim() ?? "";
    if (valid?.includes(input) ?? true) return input;
    console.log(`"${input}" is not a valid answer`);
  }
}

question("foo", ["bar", "baz"]);
